app/scripts/migrations/127.ts

```diff
@@ -0,0 +1,399 @@
+    // configuration for this chain id. The network controller requires
+    // configurations to have at least 1 endpoint.
+    if (rpcEndpoints.length === 0) {
+      return acc;
+    }
+
+    // Use the tie breaker network as the default rpc endpoint
+    const defaultRpcEndpointIndex = Math.max(
+      rpcEndpoints.findIndex(
+        (endpoint) => endpoint.networkClientId === tieBreaker?.id,
+      ),
+      // Or arbitrarily default to the first endpoint if we don't have a tie breaker
+      0,
+    );
+
+    // Calculate the unique array of non-empty block explorer urls
+    const blockExplorerUrls = [
+      ...networks.reduce((urls, network) => {
+        if (
+          isObject(network.rpcPrefs) &&
+          typeof network.rpcPrefs.blockExplorerUrl === 'string' &&
+          network.rpcPrefs.blockExplorerUrl
+        ) {
+          urls.add(network.rpcPrefs.blockExplorerUrl);
+        }
+        return urls;
+      }, new Set()),
+    ];
+
+    // Use the tie breaker network as the default block explorer, if it has one
+    const defaultBlockExplorerUrlIndex =
+      blockExplorerUrls.length === 0
+        ? undefined
+        : Math.max(
+            blockExplorerUrls.findIndex(
+              (url) =>
+                isObject(tieBreaker?.rpcPrefs) &&
+                url === tieBreaker.rpcPrefs.blockExplorerUrl,
+            ),
+            // Or arbitrarily default to the first url
+            0,
+          );
+
+    // Use the cononical network name and currency, if we have constants for them.
+    // Otherwise prefer the tie breaker's name + currency, if it defines them.
+    // Otherwise fall back to the name + currency from arbitrary networks that define them.
+    const name =
+      NETWORK_TO_NAME_MAP[chainId as keyof typeof NETWORK_TO_NAME_MAP] ??
+      tieBreaker?.nickname ??
+      networks.find((n) => n.nickname)?.nickname;
+
+    const nativeCurrency =
+      CHAIN_ID_TO_CURRENCY_SYMBOL_MAP[
+        chainId as keyof typeof CHAIN_ID_TO_CURRENCY_SYMBOL_MAP
+      ] ??
+      tieBreaker?.ticker ??
+      networks.find((n) => n.ticker)?.ticker;
+
+    acc[chainId] = {
+      chainId,
+      rpcEndpoints,
+      defaultRpcEndpointIndex,
+      blockExplorerUrls,
+      ...(defaultBlockExplorerUrlIndex !== undefined && {
+        defaultBlockExplorerUrlIndex,
+      }),
+      name,
+      nativeCurrency,
+    };
+    return acc;
+  }, {});
+
+  // Ensure that selectedNetworkClientId points to some endpoint of
+  // some network configuration. It may not, if its endpoint was not
+  // well formed or a duplicate.  In that case, fallback to mainnet.
+  const selectedNetworkClientId =
+    Object.values(networkConfigurationsByChainId)
+      .flatMap((n) =>
+        isObject(n) && Array.isArray(n.rpcEndpoints) ? n.rpcEndpoints : [],
+      )
+      .find((e) => e.networkClientId === networkState.selectedNetworkClientId)
+      ?.networkClientId ?? 'mainnet';
```

Comment:
probably too much work, but it may be possible to figure out what chainId the selectedNetworkClientId was pointing at by looking at state before we transform it? If so, it would be possible to at least resolve to the defaultRpcEndpoint for the chain?

Reply:
feel free to not do this of course. I don't think there's necessarily much gain from doing so

Reply:
providerConfig no longer exists, but yes you could peep into the old state and try to fall back to the default on the same chain instead of mainnet.  I'll consider that, although we're in edge cases of edge cases here